### PR TITLE
Interpreter: Serialize executors

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609

Problem Summary:

Currenly we don't have a serailizer of Executors that contains necessary and clean infomation.
The default serailizer of tipb::Executor contains too much info that may be hard to read.
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/TestUtils/executorSerializer.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+private:
+    void serialize(const tipb::Executor & root_executor, size_t level)
+    {
+        auto append_str = [&level, this](const tipb::Executor & executor) {
+            assert(executor.has_executor_id());
+            addPrefix(level);
+            switch (executor.tp())
+            {
+            case tipb::ExecType::TypeTableScan:
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.h:69]: and `TypePartitionTableScan`?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    case tipb::JoinType::TypeRightOuterJoin:
+        return "RightOuterJoin";
+    case tipb::JoinType::TypeLeftOuterSemiJoin:
+        return "LeftOuterSemiJoin";
+    case tipb::JoinType::TypeAntiSemiJoin:
+        return "AntiSemiJoin";
+    case tipb::JoinType::TypeInnerJoin:
+        return "InnerJoin";
+    case tipb::JoinType::TypeSemiJoin:
+        return "SemiJoin";
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:77]: throw exception in default?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    }
+}
+
+String getJoinExecTypeName(tipb::JoinExecType tp)
+{
+    switch (tp)
+    {
+    case tipb::JoinExecType::TypeHashJoin:
+        return "HashJoin";
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:86]: ditto

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+{
+    switch (tp)
+    {
+    case tipb::ExchangeType::Broadcast:
+        return "Broadcast";
+    case tipb::ExchangeType::PassThrough:
+        return "PassThrough";
+    case tipb::ExchangeType::Hash:
+        return "Hash";
+    }
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:100]: ditto

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        [&](const auto & key, FmtBuffer & fb) {
+            fb.fmtAppend("type: {}", getFieldTypeName(key.field_type().tp()));
+        },
+        ", ");
+    context.buf.append("}\n");
+}
+
+void serializeExchangeSender(const String & executor_id, const tipb::ExchangeSender & sender, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{}|type:{}\n", executor_id, getExchangeTypeName(sender.tp()));
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:287]: add the schema of ExchangeSender?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+}
+
+void serializeExchangeReceiver(const String & executor_id, const tipb::ExchangeReceiver & receiver, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{}|type:{}\n", executor_id, getExchangeTypeName(receiver.tp()));
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:292]: add the schema of ExchangeReceiver?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+            expr.children().end(),
+            [&](const auto & co, FmtBuffer &) {
+                serializeExpression(co, context);
+            },
+            ", ");
+        context.buf.append(")");
+    }
+    else
+    {
+        context.buf.fmtAppend("columnType: {}", getFieldTypeName(expr.field_type().tp()));
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:192]: how about `<column_index, column_type>`?
```
column_index = ecodeDAGInt64(expr.val())
```

File: dbms/src/TestUtils/executorSerializer.cpp (hunk 6 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/executorSerializer.cpp:292]: I will modify receiver's type to match the type of sender too

File: dbms/src/TestUtils/tests/gtest_mock_executors.cpp [hunk 8]
```diff
@@ -129,8 +136,8 @@ try
     auto request = context.scan("test_db", "test_table")
                        .aggregation(Max(col("s1")), col("s2"))
                        .build(context);
-    String expected_string = "aggregation_1\n"
-                             " table_scan_0\n";
+    String expected_string = "aggregation_1|group_by: columns:{columnType: String}, agg_func:{max(columnType: String)}\n"
```
Comment (SeaRise) [inline dbms/src/TestUtils/tests/gtest_mock_executors.cpp:139]:
```suggestion
    String expected_string = "aggregation_1 | group_by: columns:{columnType: String}, agg_func:{max(columnType: String)}\n"
```

Comment (ywqzzy) [reply dbms/src/TestUtils/tests/gtest_mock_executors.cpp:139]: I will fix all of them.

File: dbms/src/TestUtils/tests/gtest_mock_executors.cpp [hunk 9]
```diff
@@ -95,14 +99,17 @@ try
...
+                             " table_scan_0 | columns:{column_index: 0, column_type: String, column_index: 1, column_type: String}\n";
     ASSERT_DAGREQUEST_EQAUL(expected_string, request);
 
     request = context.scan("test_db", "test_table_1")
                   .limit(lit(Field(static_cast<UInt64>(10))))
                   .build(context);
-    ASSERT_DAGREQUEST_EQAUL(expected_string, request);
+
+    String expected_string_1 = "limit_1 | 10\n"
+                               " table_scan_0 | columns:{column_index: 0, column_type: Long, column_index: 1, column_type: String, column_index: 2, column_type: String}\n";
```
Comment (SeaRise) [inline dbms/src/TestUtils/tests/gtest_mock_executors.cpp:111]: how about
```suggestion
                               " table_scan_0 | columns:{index: 0, type: Long, index: 1, type: String, index: 2, type: String}\n";
```

Comment (ywqzzy) [reply dbms/src/TestUtils/tests/gtest_mock_executors.cpp]: fixed all of them

File: dbms/src/TestUtils/tests/gtest_mock_executors.cpp [hunk 10]
```diff
@@ -41,27 +41,29 @@ TEST_F(MockDAGRequestTest, MockTable)
...
-    String expected_string_1 = "table_scan_0\n";
-    ASSERT_DAGREQUEST_EQAUL(expected_string_1, request);
+    String expected_string = "table_scan_0 | columns:{index: 0, type: String, index: 1, type: String}\n";
+    ASSERT_DAGREQUEST_EQAUL(expected_string, request);
 
     request = context.scan("test_db", "test_table_1").build(context);
-    String expected_string_2 = "table_scan_0\n";
-    ASSERT_DAGREQUEST_EQAUL(expected_string_2, request);
+    String expected_string_1 = "table_scan_0 | columns:{index: 0, type: Long, index: 1, type: String, index: 2, type: String}\n";
+    ASSERT_DAGREQUEST_EQAUL(expected_string_1, request);
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/tests/gtest_mock_executors.cpp:49]:
```suggestion
    {
        String expected_string = "table_scan_0 | columns:{index: 0, type: String, index: 1, type: String}\n";
        ASSERT_DAGREQUEST_EQAUL(expected_string, request);
    }
    request = context.scan("test_db", "test_table_1").build(context);
    {
        String expected_string = "table_scan_0 | columns:{index: 0, type: Long, index: 1, type: String, index: 2, type: String}\n";
        ASSERT_DAGREQUEST_EQAUL(expected_string, request);
    }
```
How about writing in this way?

File: dbms/src/TestUtils/tests/gtest_mock_executors.cpp [hunk 11]
```diff
@@ -41,27 +41,29 @@ TEST_F(MockDAGRequestTest, MockTable)
...
     ASSERT_DAGREQUEST_EQAUL(expected_string, request);
 
     request = context.scan("test_db", "test_table_1")
-                  .filter(And(eq(col("s1"), col("s2")), lt(col("s2"), col("s3"))))
+                  .filter(And(eq(col("s1"), col("s2")), lt(col("s2"), lt(col("s1"), col("s2")))))
                   .build(context);
-    ASSERT_DAGREQUEST_EQAUL(expected_string, request);
+    String expected_string_1 = "selection_1 | equals(index: 0, type: Long, index: 1, type: String) and less(index: 1, type: String, less(index: 0, type: Long, index: 1, type: String))}\n"
+                               " table_scan_0 | columns:{index: 0, type: Long, index: 1, type: String, index: 2, type: String}\n";
+    ASSERT_DAGREQUEST_EQAUL(expected_string_1, request);
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/tests/gtest_mock_executors.cpp:66]: ditto

File: dbms/src/TestUtils/tests/gtest_mock_executors.cpp [hunk 12]
```diff
@@ -197,14 +206,14 @@ try
...
+    String expected_string = "exchange_receiver_0 | type:PassThrough, fields:{type: String, type: String, type: String}\n";
     ASSERT_DAGREQUEST_EQAUL(expected_string, request);
 
     request = context.receive("sender_1")
                   .topN("s1", false, 10)
                   .build(context);
-    String expected_string_2 = "topn_1\n"
-                               " exchange_receiver_0\n";
+    String expected_string_2 = "topn_1 | order_by: columns{index: 0, type: String, desc: true}, limit: 10\n"
+                               " exchange_receiver_0 | type:PassThrough, fields:{type: String, type: String, type: String}\n";
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/tests/gtest_mock_executors.cpp:216]: The same as before, may be it is better to write each test case in `{}`

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+        if (executor.has_join())
+        {
+            append_str(executor);
+            for (const auto & child : executor.join().children())
+                serialize(child, level);
+            return false;
+        }
+        else
+        {
+            append_str(executor);
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/executorSerializer.cpp:379]: extract `append_str(executor);` before `if-else` statement

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+String getFieldTypeName(Int32 tp)
+{
+    switch (tp)
+    {
+    case TiDB::TypeTiny:
+        return "tiny";
+    case TiDB::TypeShort:
+        return "short";
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/executorSerializer.cpp:115]: maybe `Short` and `Tiny`? And can use [M Macros](https://stackoverflow.com/a/49595815) to get enum item name from its value, just like https://github.com/pingcap/tiflash/blob/master/dbms/src/Storages/Transaction/TiDB.h#L59-L87

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+
+void serializeTableScan(const String & executor_id, const tipb::TableScan & ts, ExecutorSerializerContext & context)
+{
+    if (ts.columns_size() == 0)
+    {
+        // no column selected, must be something wrong
+        throw TiFlashException("No column is selected in table scan executor", Errors::Coprocessor::BadRequest);
+    }
+    context.buf.fmtAppend("{} | columns:{{", executor_id);
+    for (int i = 0; i < ts.columns_size(); i++)
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:147]:
```suggestion
    for (int i = 0; i < ts.columns_size(); ++i)
```

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+        agg.group_by().end(),
+        [&](const auto & by_item, FmtBuffer &) {
+            serializeExpression(by_item, context);
+        },
+        ", ");
+    context.buf.append("}, agg_func:{");
+    context.buf.joinStr(
+        agg.agg_func().begin(),
+        agg.agg_func().end(),
+        [&](const auto & by_item, FmtBuffer &) {
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:222]:
```suggestion
        [&](const auto & func, FmtBuffer &) {
```

File: dbms/src/TestUtils/executorSerializer.cpp (hunk 15 above)
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:147]: how about
```
        int i = 0;
        context.buf.joinStr(
            i,
            ts.columns_size(),
            [&](const auto & index, FmtBuffer &) {
                context.buf.fmtAppend("index: {}, type: {}", index getFieldTypeName(ts.columns().at(index).tp()));
            },
            ", ");
```

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+void serializeExchangeReceiver(const String & executor_id, const tipb::ExchangeReceiver & receiver, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | type:{}, fields:{{", executor_id, getExchangeTypeName(receiver.tp()));
+    context.buf.joinStr(
+        receiver.field_types().begin(),
+        receiver.field_types().end(),
+        [](const auto & field, FmtBuffer & fb) {
+            fb.fmtAppend("type: {}", getFieldTypeName(field.tp()));
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:284]: and column index?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+void serializeExchangeSender(const String & executor_id, const tipb::ExchangeSender & sender, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | type:{}, fields:{{", executor_id, getExchangeTypeName(sender.tp()));
+    context.buf.joinStr(
+        sender.all_field_types().begin(),
+        sender.all_field_types().end(),
+        [](const auto & field, FmtBuffer & fb) {
+            fb.fmtAppend("type: {}", getFieldTypeName(field.tp()));
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:271]: index?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+        [&](const auto & key, FmtBuffer & fb) {
+            fb.fmtAppend("type: {}", getFieldTypeName(key.field_type().tp()));
+        },
+        ", ");
+    context.buf.append("}, right_join_keys: {");
+    context.buf.joinStr(
+        join.left_join_keys().begin(),
+        join.left_join_keys().end(),
+        [&](const auto & key, FmtBuffer & fb) {
+            fb.fmtAppend("type: {}", getFieldTypeName(key.field_type().tp()));
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:258]: and index?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+        },
+        ", ");
+    context.buf.append("}, right_join_keys: {");
+    context.buf.joinStr(
+        join.left_join_keys().begin(),
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:255]: left_join_keys?
or right_join_keys?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+    {
+    case tipb::ExchangeType::Broadcast:
+        return "Broadcast";
+    case tipb::ExchangeType::PassThrough:
+        return "PassThrough";
+    case tipb::ExchangeType::Hash:
+        return "Hash";
+    default:
+        throw TiFlashException(fmt::format("Not supported Exchange type: {}", tp), Errors::Coprocessor::Internal);
+    }
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:93]: how about move this function to `DAGUtils`
and reuse in https://github.com/pingcap/tiflash/blob/master/dbms/src/Flash/Statistics/ExchangeSenderImpl.cpp#L35-L48

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+
+void serializeJoin(const String & executor_id, const tipb::Join & join, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | {}, {}. left_join_keys: {{", executor_id, getJoinTypeName(join.join_type()), getJoinExecTypeName(join.join_exec_type()));
+    for (int i = 0; i < join.left_join_keys_size(); ++i)
+    {
+        context.buf.fmtAppend("index: {}, type: {}", i, getFieldTypeName(join.left_join_keys(i).field_type().tp()));
+        if (i != join.left_join_keys_size() - 1)
+            context.buf.append(", ");
+    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/executorSerializer.cpp:159]: ditto, can avoid `if` in loop to reduce cyclomatic complexity and jump

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+    context.buf.append("}, right_join_keys: {");
+    for (int i = 0; i < join.right_join_keys_size(); ++i)
+    {
+        context.buf.fmtAppend("index: {}, type: {}", i, getFieldTypeName(join.right_join_keys(i).field_type().tp()));
+        if (i != join.right_join_keys_size() - 1)
+            context.buf.append(", ");
+    }
+    context.buf.append("}\n");
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/executorSerializer.cpp:167]: ditto

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+void serializeExchangeReceiver(const String & executor_id, const tipb::ExchangeReceiver & receiver, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | type:{}, fields:{{", executor_id, getExchangeTypeName(receiver.tp()));
+    for (int i = 0; i < receiver.field_types_size(); ++i)
+    {
+        context.buf.fmtAppend("index: {}, type: {}", i, getFieldTypeName(receiver.field_types(i).tp()));
+        if (i != receiver.field_types_size() - 1)
+            context.buf.append(", ");
+    }
+    context.buf.append("}\n");
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/executorSerializer.cpp:191]: ditto

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+void serializeExchangeSender(const String & executor_id, const tipb::ExchangeSender & sender, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | type:{}, fields:{{", executor_id, getExchangeTypeName(sender.tp()));
+    for (int i = 0; i < sender.all_field_types_size(); ++i)
+    {
+        context.buf.fmtAppend("index: {}, type: {}", i, getFieldTypeName(sender.all_field_types(i).tp()));
+        if (i != sender.all_field_types_size() - 1)
+            context.buf.append(", ");
+    }
+    context.buf.append("}\n");
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/executorSerializer.cpp:179]: ditto

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+            fb.fmtAppend(", desc: {}", order_by.desc());
+        },
+        ", ");
+    context.buf.fmtAppend("}}, limit: {}\n", top_n.limit());
+}
+
+void serializeJoin(const String & executor_id, const tipb::Join & join, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | {}, {}. left_join_keys: {{", executor_id, getJoinTypeName(join.join_type()), getJoinExecTypeName(join.join_exec_type()));
+    int bound = join.left_join_keys_size() - 1;
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:154]: assert(join.left_join_keys_size() > 0)?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 27]
```diff
@@ -0,0 +1,261 @@
...
+    for (int i = 0; i < bound; ++i)
+    {
+        context.buf.fmtAppend("index: {}, type: {}, ", i, getFieldTypeName(join.left_join_keys(i).field_type().tp()));
+    }
+    context.buf.fmtAppend("index: {}, type: {}", bound, getFieldTypeName(join.left_join_keys(bound).field_type().tp()));
+
+    context.buf.append("}, right_join_keys: {");
+    bound = join.right_join_keys_size() - 1;
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:162]: assert(join.right_join_keys_size() > 0)?

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 28]
```diff
@@ -0,0 +1,261 @@
...
+    }
+    context.buf.fmtAppend("index: {}, type: {}", bound, getFieldTypeName(join.right_join_keys(bound).field_type().tp()));
+
+    context.buf.append("}\n");
+}
+
+void serializeExchangeSender(const String & executor_id, const tipb::ExchangeSender & sender, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | type:{}, fields:{{", executor_id, getExchangeTypeName(sender.tp()));
+    int bound = sender.all_field_types_size() - 1;
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:175]: ditto

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 29]
```diff
@@ -0,0 +1,261 @@
...
+        context.buf.fmtAppend("index: {}, type: {}, ", i, getFieldTypeName(sender.all_field_types(i).tp()));
+    }
+    context.buf.fmtAppend("index: {}, type: {}", bound, getFieldTypeName(sender.all_field_types(bound).tp()));
+    context.buf.append("}\n");
+}
+
+void serializeExchangeReceiver(const String & executor_id, const tipb::ExchangeReceiver & receiver, ExecutorSerializerContext & context)
+{
+    context.buf.fmtAppend("{} | type:{}, fields:{{", executor_id, getExchangeTypeName(receiver.tp()));
+    int bound = receiver.field_types_size() - 1;
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:187]: ditto

File: dbms/src/TestUtils/executorSerializer.h [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+#include <TestUtils/TiFlashTestException.h>
+namespace DB
+{
+class Context;
+
+namespace tests
+{
+struct ExecutorSerializerContext
+{
+    Context & context;
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.h:29]: why not `FmtBuffer buf;`?
Is it necessary to pass buf in externally?

Comment (ywqzzy) [reply dbms/src/TestUtils/executorSerializer.h]: I have removed the ExecutorSerializerContext

File: dbms/src/TestUtils/mockExecutor.cpp [hunk 31]
```diff
@@ -40,7 +41,7 @@ ASTPtr buildOrderByItemList(MockOrderByItems order_by_items)
     size_t i = 0;
     for (auto item : order_by_items)
     {
-        int direction = item.second ? 1 : -1; // todo
+        int direction = item.second ? -1 : 1;
```
Comment (SeaRise) [inline dbms/src/TestUtils/mockExecutor.cpp:44]: why?
`1 : -1` change to `1 ： -1`

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 32]
```diff
@@ -0,0 +1,265 @@
...
+        ", ");
+    buf.fmtAppend("}}, limit: {}\n", top_n.limit());
+}
+
+void serializeJoin(const String & executor_id, const tipb::Join & join, FmtBuffer & buf)
+{
+    assert(join.left_join_keys_size() > 0);
+    assert(join.right_join_keys_size() > 0);
+    buf.fmtAppend("{} | {}, {}. left_join_keys: {{", executor_id, getJoinTypeName(join.join_type()), getJoinExecTypeName(join.join_exec_type()));
+    int bound = join.left_join_keys_size() - 1;
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:157]: There may be a Cartesian join?
```
join.left_join_keys_size() == join.right_join_keys_size() == 0
```

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 33]
```diff
@@ -0,0 +1,265 @@
...
+        buf.fmtAppend("<{}, {}>, ", i, getFieldTypeName(join.left_join_keys(i).field_type().tp()));
+    }
+    buf.fmtAppend("<{}, {}>", bound, getFieldTypeName(join.left_join_keys(bound).field_type().tp()));
+    buf.append("}, right_join_keys: {");
+    bound = join.right_join_keys_size() - 1;
+    for (int i = 0; i < bound; ++i)
+    {
+        buf.fmtAppend("<{}, {}>, ", i, getFieldTypeName(join.right_join_keys(i).field_type().tp()));
+    }
+    buf.fmtAppend("<{}, {}>", bound, getFieldTypeName(join.right_join_keys(bound).field_type().tp()));
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:169]: how about wrap a  function in `executorSerizlize.cpp`

File: dbms/src/TestUtils/mockExecutor.cpp (hunk 31 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/mockExecutor.cpp:44]: 1 for ASC, -1 for DESC, so the previous implementation is wrong.

File: dbms/src/TestUtils/executorSerializer.cpp (hunk 32 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/executorSerializer.cpp:157]: Currently only support inner join and outer join.

File: dbms/src/TestUtils/executorSerializer.cpp (hunk 33 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/executorSerializer.cpp:169]: Good idea! I will fix it.

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 34]
```diff
@@ -0,0 +1,276 @@
...
+    buf.fmtAppend("{} | {}, {}. left_join_keys: {{", executor_id, getJoinTypeName(join.join_type()), getJoinExecTypeName(join.join_exec_type()));
+    toString(join.left_join_keys(), buf);
+    buf.append("}, right_join_keys: {");
+    toString(join.right_join_keys(), buf);
+    buf.append("}\n");
+}
+
+void serializeExchangeSender(const String & executor_id, const tipb::ExchangeSender & sender, FmtBuffer & buf)
+{
+    assert(sender.all_field_types_size() > 0);
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:197]: nit: after wraping a function `toString`, `assert(sender.all_field_types_size() > 0);` is useless

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 35]
```diff
@@ -0,0 +1,276 @@
...
+    buf.fmtAppend("{} | type:{}, {{", executor_id, getExchangeTypeName(sender.tp()));
+    toString(sender.all_field_types(), buf);
+    buf.append("}\n");
+}
+
+void serializeExchangeReceiver(const String & executor_id, const tipb::ExchangeReceiver & receiver, FmtBuffer & buf)
+{
+    assert(receiver.field_types_size() > 0);
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:205]: ditto

File: dbms/src/TestUtils/executorSerializer.cpp [hunk 36]
```diff
@@ -0,0 +1,276 @@
...
+            fb.fmtAppend(", desc: {})", order_by.desc());
+        },
+        ", ");
+    buf.fmtAppend("}}, limit: {}\n", top_n.limit());
+}
+
+void serializeJoin(const String & executor_id, const tipb::Join & join, FmtBuffer & buf)
+{
+    assert(join.left_join_keys_size() > 0);
+    assert(join.right_join_keys_size() > 0);
```
Comment (SeaRise) [inline dbms/src/TestUtils/executorSerializer.cpp:187]: nit: after wraping a function `toString`, these codes is useless

**Review by SeaRise — APPROVED** [review]

**Pull request merged**